### PR TITLE
Floquet engine initial outline and example

### PR DESCRIPTION
**Pull request by npeard**
Basic functionality of a class that generates a class of interactions that is used by a template class to perform exact diagonalization, DMRG, etc. Currently, only implementing QuSpin exact diagonalization.

### CODE REVIEW

File: models/spin_chain.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+                              s(t, i, j)/(np.abs(i-j)**alpha), i, j]
+                             for i in range(L) for j in range(L)]
+                else:
+                    # Constant interaction strength, inverse range alpha
+                    graph = [[lambda t, s=strength:
+                              s/(np.abs(i-j)**alpha), i, j]
+                             for i in range(L) for j in range(L)]
+
+                # Add to the dictionary for this operator
+                new_interaction_dict[operator].extend(graph)
```
Comment (neutrinonerd3333) [inline models/spin_chain.py:108]: this logic looks repetitive -- probably refactorable? for example, it looks like you can handle the two ways to specify `strength` as
```python
strength_func = strength if callable(strength) else (lambda t, i, j: strength)
```

File: models/spin_chain.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        self.L = L  # number of sites
+        self.interaction_dict: Dict[str, List[List[Any]]] = interaction_dict or {}
+
+    def __call__(self, t):
+        return {op: [[f(t) if callable(f) else f for f in interaction] for
+                     interaction in interactions] for op, interactions in
+                self.interaction_dict.items()}
+
+    @classmethod
+    def from_interactions(cls, L: int, terms: List[List[Any]]):
```
Comment (neutrinonerd3333) [inline models/spin_chain.py:23]: looks like the expectation for `terms` is quite complex; probably worthwhile here to have a docstring describing parameter preconditions and expected func behavior.

also if you are into typehints: you could be super descriptive and replace `Any` with `tuple[str, StrengthSpec, RangeSpec]` with `StrengthSpec = Callable[[float, int, int], float] | float` and `RangeSpec = Literal['nn', 'nnn'] | float`, which itself would also make your intent a lot clearer :)

File: models/spin_chain.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        conjH2 = np.matrix(H2).getH()
+        product = matvec(conjH1, H2)
+        overlap = np.abs(np.trace(product))
+        return np.sqrt(overlap)
+
+    def norm_identity_loss(self, H1, H2):
+        # Use the norm difference between the product of two unitaries and the
+        # identity to establish a loss metric for the two unitaries. Identical
+        # unitaries return values close to zero.
+        # expects input of Hamiltonians times evolution time, then converts to unitaries
```
Comment (neutrinonerd3333) [inline models/spin_chain.py:153]: might as well make this a docstring

File: models/spin_chain.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+        t : float, optional
+            Time point for Hamiltonian construction
+        """
+        pass
+
+    def frobenius_loss(self, H1, H2):
+        # Use the (normalized) Frobenius norm to compute a fidelity metric for
+        # two Hamiltonians. If the two input matrices are identical,
+        # the output will be 1. Also could use numpy.linalg.norm(), which is
+        # faster?
```
Comment (neutrinonerd3333) [inline models/spin_chain.py:137]: if you want to benchmark rigorously/professionally, look into `timeit`

**Review by neutrinonerd3333 — CHANGES_REQUESTED** [review]
Solid, class hierarchy looks fine. I left mostly nitpicking comments. The only thing I would insist on is avoiding variable names that look like Python TypeVars

File: models/spin_chain.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        -----------
+        t : float, optional
+            Time point for Hamiltonian construction
+        """
+        pass
+
+    def frobenius_norm(self, H1, H2):
+        # Use the Frobenius norm to compute a fidelity metric for two
+        # Hamiltonians. If the two input matrices are identical, the output will be
+        # 1. Also could use numpy.linalg.norm(), which is faster?
```
[the remark this answers is not in the document]
Comment (npeard) [reply models/spin_chain.py]: Will do this later

File: models/spin_chain.py (hunk 1 above)
Comment (npeard) [reply models/spin_chain.py:108]: Tried it, doesn't work

File: models/spin_chain.py (hunk 2 above)
Comment (npeard) [reply models/spin_chain.py:23]: docstring and notebook example should help

**Review by neutrinonerd3333 — APPROVED** [review]

**Pull request merged**